doc/designs/quic-design/quic-api.md

```diff
@@ -818,6 +818,219 @@ reason string in bytes.
+ * any such call is a no-op.
+ *
+ * For QUIC:
+ *   Creates a new stream. Must be called only on a QUIC connection SSL object.
+ *   Can be used on client or server. If the SSL_FLAG_UNI flag is set, the
+ *   created stream is unidirectional, otherwise it is bidirectional.
+ *
+ *   To be MSMT-safe.
+ *
+ * For TLS and DTLS SSL objects:
+ *   Always fails.
+ */
+#define SSL_FLAG_UNI    1
+
+SSL *SSL_new_stream(SSL *ssl, uint64_t flags);
+```
+
+#### `SSL_accept_stream`
+
+```c
+/*
+ * Create a new SSL object representing an additional stream which has created
+ * by the peer.
+ *
+ * There is no need to call SSL_accept on the resulting object, and
+ * any such call is a no-op.
+ *
+ * For QUIC:
+ *   Must be called only on a QUIC connection SSL object. Fails if called on a
+ *   stream object. Checks if a new stream has been created by the peer. If it
+ *   has, creates a new SSL object to represent it and returns it. Otherwise,
+ *   returns NULL.
+ *
+ * For all other methods:
+ *   Returns NULL.
+ *
+ * The flags argument is unused and should be set to zero.
+ *
+ * To be MSMT-safe (i.e., can be called from multiple threads).
+ *
+ * If the QUIC connection SSL object is configured in blocking mode, this
+ * function will block unless the SSL_ACCEPT_STREAM_NO_BLOCK flag is passed.
+ */
+#define SSL_ACCEPT_STREAM_NO_BLOCK      1
+
+SSL *SSL_accept_stream(SSL *ssl, uint64_t flags);
+```
+
+#### `SSL_get_accept_stream_queue_len`
+
+```c
+/*
+ * Determine the number of streams waiting to be returned on a subsequent call
+ * to SSL_accept_stream. If this returns a non-zero value, SSL_accept_stream is
+ * guaranteed to work. Returns 0 for non-QUIC objects, or for QUIC stream
+ * objects.
+ *
+ * To be MSMT-safe.
+ */
+int SSL_get_accept_stream_queue_len(SSL *ssl);
+```
+
+#### `SSL_set_default_stream_type`
+
+```c
+/*
+ * Set a SSL_STREAM_TYPE value which determines the type of the default stream.
+ * The default value for QUIC methods is SSL_STREAM_TYPE_BIDI.
+ *
+ * This must be set before connecting. It is an error to try to set it to
+ * anything other than SSL_STREAM_TYPE_BIDI for non-QUIC methods (or
+ * we could just have this always fail on non-QUIC methods).
+ *
+ * If this is set to SSL_STREAM_TYPE_NONE, no default stream is created
+ * at connection time and the initial connection object does not represent
+ * a stream. Calls to SSL_read()/SSL_write() on this object fail. Streams
+ * must be created using SSL_new_stream() or SSL_accept_stream().
+ */
+__owur int SSL_set_default_stream_type(SSL *ssl, int type);
+```
+
+#### `SSL_detach_stream`
+
+```c
+/*
+ * Detaches a default stream from a QUIC connection object. If the
+ * QUIC connection object does not contain a default stream, returns NULL.
+ * After calling this, calling SSL_get_stream_type on the connection object
+ * returns SSL_STREAM_TYPE_NONE. Always returns NULL for non-QUIC connections.
+ */
+SSL *SSL_detach_stream(SSL *ssl);
+```
+
+#### `SSL_attach_stream`
```

Comment:
This seems out of place.  An application has to be quite aware of what's going on to make use of this function and that's not the kind of application default streams were designed for.

Reply:
Hmm, it's mainly just used for symmetry. If you can detach a stream, may as well let one be reattached. It doesn't cost us anything in terms of implementation to provide this as it's trivial. May also be useful for testing.

Reply:
Everything costs something, this might be a small cost but it will be a cost.